jerry-core/ecma/base/ecma-globals.h

```diff
@@ -196,6 +196,25 @@ typedef int32_t ecma_integer_value_t;
 typedef uintptr_t ecma_external_pointer_t;
 
 /**
+ * Representation for native handle package.
+ */
+
+typedef struct
+{
+  ecma_external_pointer_t handle_p; /**< points to the external native object */
+  ecma_external_pointer_t info_p; /**< type info of the native handle */
+} ecma_native_handle_package_t;
+
+/**
+ * Representation of the native handle package for deprecated api.
+ */
+typedef struct
+{
+  ecma_native_handle_package_t base; /**< the package's info_p is NULL */
+  ecma_external_pointer_t free_cb; /**< free callback of the native handle */
+} ecma_native_handle_package_old_t;
```

Comment:
I have been thinking about this, since it increases the memory consumption of the backward compatible variant. What about introducing two internal string types, one for the old, and another for the new API? In this case both variants would only use two pointers (8 byte on 32 bit machines), and the memory consumption of old applications would not increase. Furthermore it would be enough to use a single ecma_native_handle_package_t.

This solution would allow applications to use independent external pointers for each object, but I don't think it is a problem. Rather it is just a bad design.